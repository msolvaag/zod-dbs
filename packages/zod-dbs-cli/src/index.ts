export type { ZodDbsCliConfig } from './types.js';
export { runCli } from './cli.js';
