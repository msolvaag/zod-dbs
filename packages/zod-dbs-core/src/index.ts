export * from './utils/index.js';

export { DatabaseConnector } from './database/DatabaseConnector.js';

export type * from './types.js';
