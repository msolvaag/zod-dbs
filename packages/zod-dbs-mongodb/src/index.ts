export * from './MongoDbProvider.js';
export * from './provider.js';
