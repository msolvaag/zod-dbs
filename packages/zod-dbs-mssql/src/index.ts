export { MsSqlServerProvider } from './MsSqlServerProvider.js';
export { createClient } from './client.js';
export { createProvider } from './provider.js';
