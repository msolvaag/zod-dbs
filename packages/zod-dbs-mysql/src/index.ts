export { MySqlConnector } from './MySqlConnector.js';
export { createClient } from './client.js';
export type { Connection as Client } from 'mysql2/promise';
