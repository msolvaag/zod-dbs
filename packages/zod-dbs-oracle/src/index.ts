export { OracleProvider } from './OracleProvider.js';
export { createClient } from './client.js';
export { createProvider } from './provider.js';
