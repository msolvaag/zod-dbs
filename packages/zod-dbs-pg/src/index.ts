export { PostgreSqlConnector } from './PostgreSqlConnector.js';
export { createClient } from './client.js';
export { createConnector } from './factory.js';
export type { Client } from 'pg';
