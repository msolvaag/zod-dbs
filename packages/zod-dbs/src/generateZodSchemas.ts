import { logDebug } from 'zod-dbs-core';

import type {
  ZodDbsConfig,
  ZodDbsDbConnector,
  ZodDbsSchemaInfo,
} from 'zod-dbs-core';

import { DEFAULT_CONFIGURATION } from './constants.js';
import { generateConstantsFile } from './generate/generateConstantsFile.js';
import { generateIndexFiles } from './generate/generateIndexFiles.js';
import { generateSchemaFiles } from './generate/generateSchemaFile.js';
import { generateTypesFile } from './generate/generateTypesFile.js';
import { clearTablesDirectory } from './utils/index.js';

/**
 * Generates Zod schemas for all tables in the specified database schema.
 */
export const generateZodSchemas = async (
  dbConnector: ZodDbsDbConnector,
  config: ZodDbsConfig
): Promise<ZodDbsSchemaInfo> => {
  const generateConfig = {
    ...DEFAULT_CONFIGURATION,
    ...config,
  };

  const { connectionString, outputDir, schemaName, cleanOutput, onProgress } =
    generateConfig;

  if (cleanOutput) {
    clearTablesDirectory(outputDir);
  }

  logDebug(`Connecting to Postgres database at ${connectionString}`);

  const schema = await dbConnector.getSchemaInformation(config);

  onProgress?.('generating', { total: schema.tables.length });

  logDebug(
    `Generating zod schemas for ${schema.tables.length} tables in db schema '${schemaName}'`
  );

  for (const table of schema.tables) {
    await generateSchemaFiles(table, generateConfig);
  }

  await generateIndexFiles(schema, generateConfig);
  await generateConstantsFile(schema, generateConfig);
  await generateTypesFile(schema, generateConfig);

  onProgress?.('done');

  return schema;
};
