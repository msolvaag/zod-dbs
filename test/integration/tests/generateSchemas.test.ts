import fs from 'fs';
import path from 'path';

import { generateZodSchemas } from '../../../src/generateZodSchemas';
import {
  getClientConnectionString,
  getOutputFiles,
  outputDir,
} from '../testDbUtils';

it('generates correct zod schemas', async () => {
  const connectionString = getClientConnectionString();

  await generateZodSchemas({
    connectionString,
    outputDir,
    jsonSchemaImportLocation: '../../json',
  });

  const outputFiles = getOutputFiles();

  for (const file of outputFiles) {
    const content = fs.readFileSync(file, 'utf8');
    expect(content).toMatchSnapshot(path.relative(outputDir, file));
  }
});
