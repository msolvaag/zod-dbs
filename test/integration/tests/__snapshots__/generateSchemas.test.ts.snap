// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates correct zod schemas > constants.ts 1`] = `
"// This file is auto-generated by zod-pg
// Do not edit this file directly.

export const TABLE_POSTS = 'posts';
export const TABLE_USERS = 'users';
"
`;

exports[`generates correct zod schemas > tables/index.ts 1`] = `
"// This file is auto-generated by zod-pg
// Do not edit this file directly.

export * from './posts.js';
export * from './users.js';
"
`;

exports[`generates correct zod schemas > tables/posts.ts 1`] = `
"// This file is auto-generated by zod-pg
// Do not edit this file directly.

import { z } from 'zod';


export const POST_STATUS = ["draft","published","archived"] as const;

export const PostsTableInsertSchema = z.object({
  user_id: z.number().int().nullable().optional(),
  title: z.string(),
  content: z.string().nullable().optional(),
  published: z.boolean().nullable().optional(),
  status: z.enum(POST_STATUS).nullable().optional(),
});

export const PostsTableSchema = z.object({
  id: z.number().int(),
  user_id: z.number().int().nullable().optional().transform(val => (val === null ? undefined : val)),
  title: z.string(),
  content: z.string().nullable().optional().transform(val => (val === null ? undefined : val)),
  published: z.boolean().nullable().optional().transform(val => (val === null ? undefined : val)),
  status: z.enum(POST_STATUS).nullable().optional().transform(val => (val === null ? undefined : val)),
});

export const PostsTableUpdateSchema = PostsTableInsertSchema.partial();

export type PostRecord = z.infer<typeof PostsTableSchema>;
export type PostInsertRecord = z.input<typeof PostsTableInsertSchema>;
export type PostUpdateRecord = z.input<typeof PostsTableUpdateSchema>;
export type PostStatus = (typeof POST_STATUS)[number];"
`;

exports[`generates correct zod schemas > tables/users.ts 1`] = `
"// This file is auto-generated by zod-pg
// Do not edit this file directly.

import { z } from 'zod';
import * as json from '../../json.js';


export const UsersTableInsertSchema = z.object({
  name: z.string().max(100),
  email: z.string().max(255).nullable().optional(),
  status: z.string().max(20).nullable().optional(),
  profile: json.UserProfileSchema.nullable().optional().transform(val => val ? JSON.stringify(val) : null),
  created_at: z.coerce.date().nullable().optional(),
});

export const UsersTableSchema = z.object({
  id: z.number().int(),
  name: z.string(),
  email: z.string().nullable().optional().transform(val => (val === null ? undefined : val)),
  status: z.string().nullable().optional().transform(val => (val === null ? undefined : val)),
  profile: json.UserProfileSchema.nullable().optional().transform(val => (val === null ? undefined : val)),
  created_at: z.coerce.date().nullable().optional().transform(val => (val === null ? undefined : val)),
});

export const UsersTableUpdateSchema = UsersTableInsertSchema.partial();

export type UserRecord = z.infer<typeof UsersTableSchema>;
export type UserInsertRecord = z.input<typeof UsersTableInsertSchema>;
export type UserUpdateRecord = z.input<typeof UsersTableUpdateSchema>;
"
`;

exports[`generates correct zod schemas > types.ts 1`] = `
"// This file is auto-generated by zod-pg
// Do not edit this file directly.

export type Table = 
  | 'posts'
  | 'users';
"
`;
